lib/engine/src/export.rs

```diff
@@ -44,17 +46,57 @@ impl From<VMExport> for Export {
     }
 }
 
+/// Extra metadata about `ExportFunction`s.
+///
+/// The metadata acts as a kind of manual virtual dispatch. We store the
+/// user-supplied `WasmerEnv` as a void pointer and have methods on it
+/// that have been adapted to accept a void pointer.
+#[derive(Debug, PartialEq)]
+pub struct ExportFunctionMetadata {
+    /// This field is stored here to be accessible by `Drop`.
+    ///
+    /// At the time it was added, it's not accessed anywhere outside of
+    /// the `Drop` implementation. This field is the "master copy" of the env,
+    /// that is, the original env passed in by the user. Every time we create
+    /// an `Instance` we clone this with the `host_env_clone_fn` field.
+    ///
+    /// Thus, we only bother to store the master copy at all here so that
+    /// we can free it.
+    ///
+    /// See `wasmer_vm::export::VMExportFunction::vmctx` for the version of
+    /// this pointer that is used by the VM when creating an `Instance`.
+    pub host_env: *mut std::ffi::c_void,
+    /// Function pointer to `WasmerEnv::init_with_instance(&mut self, instance: &Instance)`.
+    ///
+    /// This function is called to finish setting up the environment after
+    /// we create the `api::Instance`.
+    // This one is optional for now because dynamic host envs need the rest
+    // of this without the init fn
+    pub import_init_function_ptr: Option<ImportInitializerFuncPtr>,
+    /// TODO:
+    pub host_env_clone_fn: fn(*mut std::ffi::c_void) -> *mut std::ffi::c_void,
+    /// The destructor to free the host environment.
+    pub host_env_drop_fn: fn(*mut std::ffi::c_void),
+}
+
+// We have to free `host_env` here because we always clone it before using it
+// so all the `host_env`s freed at the `Instance` level won't touch the original.
+impl Drop for ExportFunctionMetadata {
+    fn drop(&mut self) {
+        if !self.host_env.is_null() {
+            (self.host_env_drop_fn)(self.host_env);
+        }
+    }
+}
+
 /// A function export value with an extra function pointer to initialize
 /// host environments.
 #[derive(Debug, Clone, PartialEq)]
 pub struct ExportFunction {
     /// The VM function, containing most of the data.
     pub vm_function: VMExportFunction,
-    /// Function pointer to `WasmerEnv::init_with_instance(&mut self, instance: &Instance)`.
-    ///
-    /// This function is called to finish setting up the environment after
-    /// we create the `api::Instance`.
-    pub import_init_function_ptr: Option<ImportInitializerFuncPtr>,
+    /// TODO:
```

Comment:
What's left to do?

Reply:
Oh documentation, we have some lints which prevent the code from compiling without doc comments everywhere so I just write `TODO` when I don't want to write doc-comments while still developing things.  I'll fill them out now though